true up c10::FunctionSchema::operator<< to print native_functions.yaml syntax

bdhirsh opened the pull request:
Specifically, `int[n]` default values are specified in `native_functions.yaml` by a single integer that then fills the whole array, while printing a `FunctionSchema` object will print out e.g. `[1, 1]` instead of 1.

True-ing this up will let executorch faithfully re-use `torchgen` infra to parse function schema objects from `torch.ops.aten.<foo>._schema`.

This is a one off fix, but it comes up again we should consider doing what Ed said and just directly plumbing the schema string into the `FunctionSchema` object.

Before:
```
>>> print(torch.ops.aten.conv2d.default._schema)
aten::conv2d(Tensor input, Tensor weight, Tensor? bias=None, int[2] stride=[1, 1], int[2] padding=[0, 0], int[2] dilation=[1, 1], int groups=1) -> Tensor
```
After:
```
>>> print(torch.ops.aten.conv2d.default._schema)
aten::conv2d(Tensor input, Tensor weight, Tensor? bias=None, int[2] stride=1, int[2] padding=0, int[2] dilation=1, int groups=1) -> Tensor
```

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #79645

bdhirsh committed:
true up c10::FunctionSchema::operator<< to print native_functions.yaml syntax

[ghstack-poisoned]

shunting314 commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -509,6 +509,18 @@ inline std::ostream& operator<<(std::ostream& out, const Argument& arg) {
+      // instead of
+      //   int[2] stride=[1, 1]
+      auto default_val = arg.default_value().value().toIntList();
+      if (default_val.size() > 1) {
```
Will there be cases that the items in `default_val` are different (e.g. [1, 3, 2] rather than [1, 1, 1])? Should we either check that or assert all items should be equal ?

bdhirsh replied:
good point, ill guard for that case

shunting314 reviewed (approved):
Thanks for the quick fix!

Leave one comment for adding some additional check. Otherwise LGTM!

bdhirsh committed:
Update on "true up c10::FunctionSchema::operator<< to print native_functions.yaml syntax"

Specifically, `int[n]` default values are specified in `native_functions.yaml` by a single integer that then fills the whole array, while printing a `FunctionSchema` object will print out e.g. `[1, 1]` instead of 1.

True-ing this up will let executorch faithfully re-use `torchgen` infra to parse function schema objects from `torch.ops.aten.<foo>._schema`.

This is a one off fix, but it comes up again we should consider doing what Ed said and just directly plumbing the schema string into the `FunctionSchema` object.

Before:
```
>>> print(torch.ops.aten.conv2d.default._schema)
aten::conv2d(Tensor input, Tensor weight, Tensor? bias=None, int[2] stride=[1, 1], int[2] padding=[0, 0], int[2] dilation=[1, 1], int groups=1) -> Tensor
```
After:
```
>>> print(torch.ops.aten.conv2d.default._schema)
aten::conv2d(Tensor input, Tensor weight, Tensor? bias=None, int[2] stride=1, int[2] padding=0, int[2] dilation=1, int groups=1) -> Tensor
```




[ghstack-poisoned]

malfet commented:
This breaks internal build with the following error/warning:
```
aten/src/ATen/core/function_schema.h:521:26: error: loop variable 'i' binds to a temporary value produced by a range of type 'integer_range<unsigned long>' [-Werror,-Wrange-loop-bind-reference]
        for (const auto& i : c10::irange(1, default_val.size())) {
```
Forward fix: https://github.com/pytorch/pytorch/pull/79815
